packages/apollo-server-lambda/src/ApolloServer.ts

```diff
@@ -59,16 +59,22 @@ function maybeCallbackify(
     event: APIGatewayProxyEvent,
     context: LambdaContext,
   ) => Promise<APIGatewayProxyResult>,
-): (
-  event: APIGatewayProxyEvent,
-  context: LambdaContext,
-  callback: APIGatewayProxyCallback | undefined,
 ) => void | Promise<APIGatewayProxyResult> {
```

Comment:
I'm not sure I understand what your PR accomplishes. Isn't the type of the return value given when we declare the function's own type here?

Can you explain what doesn't work about the current code and how your change improves it? I don't doubt that you've got a good idea here — I just can't figure out what it is.

Reply:
Today, the handler returned by `apollo-server-lambda` returns `void | Promise<APIGatewayProxyResult>` but we can do better. We can use multiple signatures to return either void (if a callback is provided) or a promise if no callback is provided. It's in the case the callback is not used directly by AWS but included in some other code.

Reply:
Hmm, I get that that is a useful goal, but how does your PR achieve that goal given that `maybeCallbackify` has an explicit return type declared?

Can you share some code that works with this change but not without this change (perhaps even as an addition to the test suite, but as a comment if that's too hard)?

Also CI says this doesn't build. Perhaps this is related to my confusion.

Reply:
OK, that makes more sense. But I worry that later refactorings could break this change since it's dependent on type inference of this function's return value. Can you write some TS code in `packages/apollo-server-lambda/src/__tests__/` that depends on this typing improvement? It doesn't really even have to test behavior, just the typing improvements you're going for here. (Though... I seem to recall that our test suites may be in bad shape where they are not fully type-checked, which might make this a bit of a challenge to prove. But maybe you can at least share what the code that only typechecks with this change would look like?)

Reply:
I don't know how to implement a test for that but here is some code (which looks like my app) that will raise a Typescript error:

```typescript
const server = new ApolloServer({
	schema: someSchema,
});

const apolloGraphqlHandler = server.createHandler();


export const handler: Handler<APIGatewayProxyEvent, APIGatewayProxyResult> = async (event, context, callback) => {
	// Do something such as loading config async

	const res = await apolloGraphqlHandler(event, context, undefined); // res can be undefined but will not since callback argument is undefined

	// Do something such as waiting for some promises

	return res; // Can not return undefined => Typescript error
};

```